crates/libsolv_rs/src/solver/mod.rs

```diff
@@ -0,0 +1,1399 @@
+    fn resolve_dependencies(&mut self, mut level: u32) -> Result<u32, Problem> {
+        let mut i = 0;
+        loop {
+            if i >= self.clauses.len() {
+                break;
+            }
+
+            let (required_by, candidate) = {
+                let clause = &self.clauses[i];
+                i += 1;
+
+                // We are only interested in requires clauses
+                let Clause::Requires(solvable_id, deps) = clause.kind else {
+                    continue;
+                };
+
+                // Consider only clauses in which we have decided to install the solvable
+                if self.decision_tracker.assigned_value(solvable_id) != Some(true) {
+                    continue;
+                }
+
+                // Consider only clauses in which no candidates have been installed
+                let candidates = &self.pool.match_spec_to_candidates[deps];
+                if candidates
+                    .iter()
+                    .any(|&c| self.decision_tracker.assigned_value(c) == Some(true))
+                {
+                    continue;
+                }
+
+                // Get the first candidate that is undecided and should be installed
+                //
+                // This assumes that the packages have been provided in the right order when the solvables were created
+                // (most recent packages first)
+                (
+                    solvable_id,
+                    candidates
+                        .iter()
+                        .cloned()
+                        .find(|&c| self.decision_tracker.assigned_value(c).is_none())
+                        .unwrap(),
+                )
+            };
+
+            level = self.set_propagate_learn(level, candidate, required_by, ClauseId::new(i))?;
+
+            // We have made progress, and should look at all clauses in the next iteration
+            i = 0;
+        }
+
+        // We just went through all clauses and there are no choices left to be made
+        Ok(level)
+    }
+
+    /// Executes one iteration of the CDCL loop
+    ///
+    /// A set-propagate-learn round is always initiated by a requirement clause (i.e.
+    /// [`Clause::Requires`]). The parameters include the variable associated to the candidate for the
+    /// dependency (`solvable`), the package that originates the dependency (`required_by`), and the
+    /// id of the requires clause (`clause_id`).
+    ///
+    /// Refer to the documentation of [`Solver::run_sat`] for details on the CDCL algorithm.
+    ///
+    /// Returns the new level after this set-propagate-learn round, or a [`Problem`] if we
+    /// discovered that the requested jobs are unsatisfiable.
+    fn set_propagate_learn(
+        &mut self,
+        mut level: u32,
+        solvable: SolvableId,
+        required_by: SolvableId,
+        clause_id: ClauseId,
+    ) -> Result<u32, Problem> {
+        level += 1;
+
+        println!(
```

Comment:
Should we move to `tracing::info!` here? Or `debug!`.

Reply:
I'm postponing that because it's a fairly mechanical change and I want to make sure all the truly important stuff is done first